### DISCUSSION TITLE
Adding aditional character set support

### DISCUSSION THREAD

**Discussion by paulGeoghegan** [nvaccess/nvda#16720] [General]
I know this isn't technically the right place but I already asked on one of the mailing lists and they said they just dealt with translating the UI for NVDA.

We have added multiple languages to Liblouis but they aren't currently supported in NVDA. Currently we have unoficially added them by updating the locale/en/symbols.dic file however this is obviously not official. Is there a way to create other directories in this folder in order to officially add them to NVDA. For example one of the languages is Akkadian which has the language code akk so could I create an akk folder instead of the en folder and then add the required files?
If I did that would I then be able to open a PR to get this officially added to NVDA?
Thanks!

**Comment by seanbudd** [comment]
See our docs on how to translate the symbols file: https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/developerGuide/developerGuide.md#symbol-pronunciation-symbolpronunciation

**Reply by CyrilleB79** [reply]
@seanbudd the indications on how translation works may be useful.

You write in https://github.com/nvaccess/nvda/pull/16732#issuecomment-2187677432,that new files should be added through the translation framework. In the current case, where @paulGeoghegan wants to add only symbol and character files, this does not seem to be suitable though. Indeed in the translation framework the interface (nvda.po) is added for each new language. In the case of ancient languages, translating the UI is useless: who wants NVDA's UI in translated cuneiform?!
In the past, for some Indian languages, the UI was not translated, probably because there is no maintainer for such language and Indian people were able to use another language for the GUI; see commit 6e42c0ccf51b5c3a60e7fa6fbeafc785ae64499f / issue #4410.
To summarize, a difference should be done between language of the UI and language of TTS or braille.

@paulGeoghegan, as pointed in #16732, it would be useful if you are trying to acheive in an issue. Do you want to improve speech (voice) reading, with which TTS? Or braille reading? Or both? Looking at some changes in #16732, it's not clear to me what it improves, hence the need to discuss first the issue and then elaborate a solution.

**Comment by paulGeoghegan** [comment]
@seanbudd we just opened #16732

**Comment by paulGeoghegan** [comment]
@seanbudd as @CyrilleB79 mentioned we don't really want to translate the UI or have a synthesizer to be able to read the language. We just want to be able to get speach for the characters and we opened a PR because it seemed like you had said all you were going to say about what I had asked so we went ahead and did what it seemed to suggest in the documentation you had sent.
The languages we added are currently available in Liblouis but not in NVDA so you can get braille output but not speach. These languages are mainly used for academic purposes.

**Reply by CyrilleB79** [reply]
@paulGeoghegan you write:
> we don't really want to (...)  have a synthesizer to be able to read the language. We just want to be able to get speach for the characters

So if you do not want to have something able to read the language but still want something to be spoken, you should really describe with more details and with an example the result that you are expecting. The use case is quite specific so it should be described with more care than you did so that it is understood.

Also please clarify how you instruct NVDA to read these languages if there is no supported TTS for them.

**Reply by paulGeoghegan** [reply]
@yeatersink talks more about this but currently we add the characters to the English locale files in order to get speech output for the characters.

**Reply by CyrilleB79** [reply]
@paulGeoghegan writes:
> @yeatersink talks more about this but currently we add the characters to the English locale files in order to get speech output for the characters.

Sorry, this is not a description of the output you get. It's a description of how you manage to get it.

I'd like a detailed description of the speech output you get. E.g. are the words spelled out instead of being spoken? Or something else?

**Reply by paulGeoghegan** [reply]
We get speech output for each character. As you can see in our PR we add the character descriptions and then names for all of the multi character combinations so it reads the names of the characters which is what it should do.
For the characters in the symbols.dic file it will read out the name we have given the character combination but not all languages have the symbols.dic file as it is not required. For example in the Akkadian languages there are no multi-character combinations or punctuation definitions.

**Comment by yeatersink** [comment]
Greetings, 
I opened the pull request because we were under the assumption that we did all that  @seanbudd suggested. 

@paulGeogheganand @CyrilleB79   is correct in that with ancient languages, we only want the character and symbol as opposed to attempting to speak the entire word or phrase. Attempting to have a synthesizer speak the entire language is very problematic for many reasons.  

Currently, we are utilizing the synthesizers provided with NVDA. These work great and offer the access with speech that is necessary. 

I am currently in a PhD program and access to these languages are necessary part of my education, and there are several more blind users who need this same level of access to these languages as part of their education as well.

Adding speech support for these languages are a bit of a breakthrough academically, and this will hopefully drive many more users to NVDA as a result of NV Access forward thinking in allowing access to languages like these and other languages as well.  

What is the next step in moving forward with this project?

**Reply by CyrilleB79** [reply]
@yeatersink, I have posted my message without having seen yours before.

I understand that you have currently a setup that meets your needs. As written before, we are not able to figure your experience.
* With your "hack", are the words spelled out in these languages? If no, could you explain how they are spoken.
* At NVDA startup, it speaks English (or your Windows language). With the build of PR #16732, how do you instruct NVDA to use transliterated cuneiform for speech?

**Comment by yeatersink** [comment]
@CyrilleB79 For our testing on our local computer, we wrote the characters for the languages to the “English” local folder. ThusNVDA utilizes the TTS that support English to read the characters.

We have the name of character being spoken, because we named that character  in English. Then it reads characters as opposed to words. For example:
A pi lum is a word which is comprised of three Akkadian characters.  A pi lum, Is the word Awilum. Which means “son” in Akkadian.  

What we want is to be able to read the characters a- pi- lum as opposed to attempting to speak the word. 

 When @seanbudd  responded the other day to @paulGeoghegan's  question, we followed the documentation instructions and created their own Locale folders for each individual languages. 

After reading your questions, we may need to do more work to ensure that the TTS of NVDA will actually read the characters. 

What would be optimal is that any NVDA user would be able to read these characters with their own default TTS while normally using NVDA. 

What do we need to do to accomplish this?

**Comment by LeonarddeR** [comment]
As I'm going to study ancient Hebrew and Greek in the near future, I'm following this discussion with interest. I think there is some confusion here. If I understand @paulGeoghegan and @yeatersink correctly, they just want their NVDA to pronounce these foreign symbols. That's also what I'd like. I don't thing that requires extra locales being added to NVDA, in fact as they already pointed out it is enough to add the several symbols to the English locale and they will be spoken, at least when navigating character by character. Therefore this request doesn't differ that much from issues like https://github.com/nvaccess/nvda/issues/5194, for example.
I don't understand why character descriptions need to be added though. I'd say defining symbols should be enough.

**Reply by paulGeoghegan** [reply]
You are correct. Given what the documentation said I thought we would need to create locales for each language and I didn't think it would be good to add thousands of characters to the English locale files but perhaps I'm wrong.
I also thought that we should define the single characters in the characterDescriptions.dic file since they are single characters but again this could also be incorrect.

**Reply by LeonarddeR** [reply]
The character descriptions are used for delayed character descriptions and when describing a character with the report character command. They are usually used to provide the name of the character as in the phonetic alphabet, e.g. Alpha, Bravo, Charli, etc. I don't think that's necessary here.

> I didn't think it would be good to add thousands of characters to the English locale files.

I share that concern, but every time this concern is raised, it results in stalled requests🙈. That said, Espeak has several of these upstream so I think we should reconsider this.

**Reply by paulGeoghegan** [reply]
Okay, well what do you suggest? I'm open to ideas and alternatives. We didn't really want to add it to Espeak, because that would limit it to only Espeak and I'm not sure how many people actually use as their synthesizer for NVDA.

**Reply by CyrilleB79** [reply]
Do you (@paulGeoghegan, @yeatersink and @LeonarddeR) want to be able to read text normally, or navigating character by character / pressing spelling commands is enough?
If you wish to read ancient textes pressing say all (NVDA+downArrow), a specific mode should be implemented to spell out any content passed to speech.

More generally, adding new characters in the symbol file works if you put level=`char` and preserve=`norep`. For example that's done in the French file for characters that are not pronounced correctly by some synths. Using the `char` level allows to have characters reported correctly when spelling out words without degrading other usages of this character.

For example, if the content of the file `source/locale/Transliterated-Cuneiform/symbols.dic` in PR #16732 was added to English symbol file, it would clearly degrade other usage of the characters listed in this file, e.g. reading French or Spanish, since it uses level=`most` and preserve=`always`.

**Reply by paulGeoghegan** [reply]
@CyrilleB79 yes we just need character by character but it would be good if it could read them without you having to navigate to each character. As in if there was a line of English text with some Akkadian characters in the middle or just a line of Akkadian characters then we would want it to read the line of characters but only the names of them.

**Reply by CyrilleB79** [reply]
Also, now that the use case has been explained (or is about to be), @LeonarddeR would you mind open an issue so that it is triaged? You are surely much more accustomated to writing issue or feature request descriptions with an accurate description of the use case. This would allow this issue to be more likely triaged by NV Access.

**Reply by LeonarddeR** [reply]
I created https://github.com/nvaccess/nvda/issues/16739 as a proposal to continue this.
